File: PriorityQueue/PriorityQueue.swift
import Foundation

class PriorityQueue {
    private var maxSize: Int
    private var queArray: [Int64]
    private var nItems: Int

    // Constructor
    init(s: Int) {
        maxSize = s
        queArray = Array(repeating: 0, count: maxSize)
        nItems = 0
    }

    // Insert an element into the priority queue
    func insert(item: Int64) {
        var j: Int // Declare the variable j here
        if nItems == 0 {                // If the queue is empty
            queArray[nItems] = item     // Insert the item into cell 0
            nItems += 1
        } else {                       // Else, the queue is not empty
            j = nItems - 1 // Initialize j outside the for loop
            for j in stride(from: j, through: 0, by: -1) { // Reverse iteration
                if item > queArray[j] {       // If the new element is greater
                    queArray[j + 1] = queArray[j] // Move the current element up
                } else {    // If the new element is less or equal
                    break     // Shift stops
                }
            }
            queArray[j + 1] = item // Insert the element
            nItems += 1
        }
    }

    // Remove the minimum element from the priority queue
    func remove() -> Int64 {
        nItems -= 1
        return queArray[nItems]
    }

    // Peek the minimum element in the priority queue
    func peekMin() -> Int64 {
        return queArray[nItems - 1]
    }

    // Check if the queue is empty
    func isEmpty() -> Bool {
        return nItems == 0
    }

    // Check if the queue is full
    func isFull() -> Bool {
        return nItems == maxSize
    }
}

// Entry point of the application
func main() {
    let thePQ = PriorityQueue(s: 5)
    thePQ.insert(item: 30)
    thePQ.insert(item: 50)
    thePQ.insert(item: 10)
    thePQ.insert(item: 40)
    thePQ.insert(item: 20)

    while !thePQ.isEmpty() {
        let item = thePQ.remove()
        print(item, terminator: " ")
    }
    print("")
}

// Call the entry point
main()

File: PriorityQueue/PriorityQueue.go
package main

import "fmt"

type PriorityQueue struct {
	maxSize  int
	queArray []int64
	nItems   int
}

// Constructor
func NewPriorityQueue(s int) *PriorityQueue {
	return &PriorityQueue{
		maxSize:  s,
		queArray: make([]int64, s),
		nItems:   0,
	}
}

// Insert an element into the priority queue
func (pq *PriorityQueue) Insert(item int64) {
	var j int
	if pq.nItems == 0 { // If the queue is empty
		pq.queArray[pq.nItems] = item // Insert the item into cell 0
		pq.nItems++
	} else { // Else, the queue is not empty
		for j = pq.nItems - 1; j >= 0; j-- { // Reverse iteration
			if item > pq.queArray[j] { // If the new element is greater
				pq.queArray[j+1] = pq.queArray[j] // Move the current element up
			} else { // If the new element is less or equal
				break // Shift stops
			}
		}
		pq.queArray[j+1] = item // Insert the element
		pq.nItems++
	}
}

// Remove the minimum element from the priority queue
func (pq *PriorityQueue) Remove() int64 {
	pq.nItems--
	return pq.queArray[pq.nItems]
}

// Peek the minimum element in the priority queue
func (pq *PriorityQueue) PeekMin() int64 {
	return pq.queArray[pq.nItems-1]
}

// Check if the queue is empty
func (pq *PriorityQueue) IsEmpty() bool {
	return pq.nItems == 0
}

// Check if the queue is full
func (pq *PriorityQueue) IsFull() bool {
	return pq.nItems == pq.maxSize
}

func main() {
	thePQ := NewPriorityQueue(5)
	thePQ.Insert(30)
	thePQ.Insert(50)
	thePQ.Insert(10)
	thePQ.Insert(40)
	thePQ.Insert(20)

	for !thePQ.IsEmpty() {
		item := thePQ.Remove()
		fmt.Print(item, " ")
	}
	fmt.Println()
}
